src/test/rpc/ShardArchiveHandler_test.cpp

```diff
@@ -49,196 +49,207 @@ class ShardArchiveHandler_test : public beast::unit_test::suite
-            auto host = server.local_endpoint().address().to_string();
-            auto port = std::to_string(server.local_endpoint().port());
-            server.stop();
+        for (int i = 0; i < env.app().getShardStore()->ledgersPerShard() *
+                 (numberOfDownloads + 1);
+             ++i)
+        {
+            env.close();
+        }
 
-            Downloads const dl = [&host, &port] {
-                Downloads ret;
+        auto handler = env.app().getShardArchiveHandler();
+        BEAST_EXPECT(handler);
+        BEAST_EXPECT(dynamic_cast<RPC::RecoveryHandler*>(handler) == nullptr);
 
-                for (int i = 1; i <= 10; ++i)
-                {
-                    ret.push_back(
-                        {i,
-                         (boost::format("https://%s:%d/%d.tar.lz4") % host %
-                          port % i)
-                             .str()});
-                }
+        auto server = createServer(env);
+        auto host = server.local_endpoint().address().to_string();
+        auto port = std::to_string(server.local_endpoint().port());
+        server.stop();
 
-                return ret;
-            }();
+        Downloads const dl = [count = numberOfDownloads, &host, &port] {
+            Downloads ret;
 
-            for (auto const& entry : dl)
+            for (int i = 1; i <= count; ++i)
             {
-                parsedURL url;
-                parseUrl(url, entry.second);
-                handler->add(entry.first, {url, entry.second});
+                ret.push_back(
+                    {i,
+                     (boost::format("https://%s:%d/%d.tar.lz4") % host % port %
+                      i)
+                         .str()});
             }
 
-            BEAST_EXPECT(handler->start());
+            return ret;
+        }();
 
-            auto stateDir = RPC::ShardArchiveHandler::getDownloadDirectory(
-                env.app().config());
+        for (auto const& entry : dl)
+        {
+            parsedURL url;
+            parseUrl(url, entry.second);
+            handler->add(entry.first, {url, entry.second});
+        }
 
-            std::unique_lock<std::mutex> lock(handler->m_);
+        BEAST_EXPECT(handler->start());
 
-            BEAST_EXPECT(
-                boost::filesystem::exists(stateDir) ||
-                handler->archives_.empty());
+        auto stateDir =
+            RPC::ShardArchiveHandler::getDownloadDirectory(env.app().config());
 
-            while (!handler->archives_.empty())
+        std::unique_lock<std::mutex> lock(handler->m_);
+
+        BEAST_EXPECT(
+            boost::filesystem::exists(stateDir) || handler->archives_.empty());
+
+        using namespace std::chrono_literals;
+        auto waitMax = 5000ms;
+
+        while (!handler->archives_.empty())
+        {
+            lock.unlock();
+            std::this_thread::sleep_for(100ms);
+
+            if (waitMax -= 100ms; waitMax == 0ms)
```

Comment:
This works as long as `waitMax` is initialized to a multiple of 100ms.  Consider testing for `waitMax <= 0ms` to make the condition testing a little less fragile.

Reply:
Done